[CP] Add assertion for unsupported load-balance + non-causal

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141622

We actually do not support load-balance mode when non_causal = True, due
to changes in data shuffling for load_balance mode.  This PR just adds
an assertion to make this limitation clear.

Fixes #141429

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

wconstab committed:
[CP] Fix logic for is_causal=False and partial update

(WIP: verified this literally fixes the issue in #141429 but not sure if
it is complete or correct). cc @fegin

Make logic in one part of backward consistent with logic in the other
part of backward, namely, if `not is_causal`, do not do partial update.

Fixes #141429

[ghstack-poisoned]

XilunWu commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -717,7 +717,7 @@ def _templated_ring_attention_backward(
             )
 
-            if i <= rank and _cp_options.enable_load_balance:
+            if i <= rank and _cp_options.enable_load_balance and is_causal:
```
I'm a bit confused on `enable_load_balance` and `is_causal`. My understanding is we only need `enable_load_balance == True` if `is_causal == True`. Then here we only need to check `enable_load_balance`.

XilunWu commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -742,7 +742,7 @@ def _templated_ring_attention_backward(
         dkv_rotater.exchange_buffers(next_grad_kv)
 
-        if i <= rank or not _cp_options.enable_load_balance:
+        if i <= rank or not _cp_options.enable_load_balance or not is_causal:
```
same as above, `not is_causal` implies `not enable_load_balance`. We don't need to check it here.

wconstab replied:
i think we may have an implicit assumption that `not is_causal` implies `not enable_load_balance`.  But we do not assert this anywhere and users are free to specify `not is_causal` and `enable_load_balance`.

@fegin was just telling me, we should probably just ban this combination as an invalid combination.

XilunWu reviewed (commented):
I believe those extra checks are not needed, but I totally could be wrong. Want to discuss to clarify.

wconstab committed:
Update on "[CP] Fix logic for is_causal=False and partial update"


Make logic in one part of backward consistent with logic in the other
part of backward, namely, if `not is_causal`, do not do partial update.

Fixes #141429

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o

[ghstack-poisoned]

wconstab committed:
Update on "[CP] Add assertion for unsupported load-balance + non-causal"


We actually do not support load-balance mode when non_causal = True, due
to changes in data shuffling for load_balance mode.  This PR just adds
an assertion to make this limitation clear.

Fixes #141429

[ghstack-poisoned]

XilunWu reviewed (approved):
Thanks for adding the validity check!

wconstab commented:
@pytorchbot merge
